File: Python/백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
# 12:30 ~ 12:57
import sys

sys.setrecursionlimit(100_000)
input = sys.stdin.readline
n, v = map(int, input().split())

visited = [0] * n
adj = [[0] * n for _ in range(n)]

for _ in range(v):
    s1, s2 = map(int, input().split())
    s1 -= 1
    s2 -= 1
    adj[s1][s2] = adj[s2][s1] = 1

def dfs(start):
    for idx in range(len(adj[start])):
        if adj[start][idx] == 1 and visited[idx] == 0:
            visited[idx] = 1
            dfs(idx)

answer = 0
for i in range(n):
    if visited[i] == 0:
        visited[i] = 1
        answer += 1
        dfs(i)

print(answer)

File: Java/백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.*;
import java.util.StringTokenizer;
import java.util.ArrayList;

public class Main {
    private static ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
    private static boolean[] visited;
    private static int answer = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        visited = new boolean[n + 1];
        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            adj.get(start).add(end);
            adj.get(end).add(start);
        }

        for (int i = 1; i <= n; i++) {
            if (visited[i] == false) {
                dfs(i);
                answer++;
            }
        }

        bw.write(Integer.toString(answer));
        bw.flush();
        br.close();
        bw.close();
    }

    private static void dfs(int node) {
        visited[node] = true;
        for (int nxtNode : adj.get(node)) {
            if (visited[nxtNode] == false) {
                dfs(nxtNode);
            }
        }
    }
}